### PR TITLE
Refactor tokenizer to avoid off by one error

### PR DESCRIPTION
**Pull request by vadim-moz**

### CODE REVIEW

File: src/tokenizers/strspn.h [hunk 1]
```diff
@@ -4,17 +4,28 @@
...
+         * Note that for valid token length is never 0.
          */
-        const char* operator()(const char* last) const {
-            size_t s = strspn(last,
-                "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
-            return (*last != '\0') ? last + s : NULL;
+        void operator()(const char* string, StringToken& output) const {
+            const char* start = strpbrk(string, alphas);
+            output.start = start;
+            output.length = (start? strspn(start, alphas): 0);
```
Comment (b4hand) [inline src/tokenizers/strspn.h]: Add space before `?` and `:` when used as ternary operator.

Comment (vadim-moz) [inline src/tokenizers/strspn.h]: ok. Is there any C++ style guide? I feel I violate a lot of recommentations...

File: src/tokenizers_test.cpp [hunk 2]
```diff
@@ -1,43 +1,73 @@
 #include "catch.hpp"
 
+#include <stdio.h>
```
Comment (b4hand) [inline src/tokenizers_test.cpp]: Was this import for debugging? You don't appear to be using it anywhere.

File: src/tokenizers_test.cpp [hunk 3]
```diff
@@ -1,43 +1,73 @@
...
+    Simhash::Strspn tokenizer;
+    Simhash::StringToken token;
+
+    const char** expected = expected_tokens;
+    for (tokenizer(string, token)
+        ; token.length != 0
+        ; tokenizer(token.start + token.length, token)) {
+        REQUIRE(((*expected) != NULL));
+        REQUIRE((token.start != NULL));
+        REQUIRE((token.length == strlen(*expected)));
```
Comment (b4hand) [inline src/tokenizers_test.cpp]: I would add a `#include <string.h>` for using `strlen` and `strncmp`.

**Comment by b4hand** [comment]
Which test was previously failing that you added?

File: src/tokenizers/strspn.h (hunk 1 above)
Comment (b4hand) [inline src/tokenizers/strspn.h]: Martin's been working on configuring a style rule checker for us, but I don't think it's ready at this point.

**Comment by vadim-moz** [comment]
The last one. The problem was not so much in tokenizer but in the `hash`'s loop. However, it _was_ difficult to write that loop clear and efficiently. Hopefully, use of `strpbrk` will increase performance as well -- although I'd replace both it and `strspn` completely (if the tokenizer is of any longterm / production value)

File: src/tokenizers/strspn.h (hunk 1 above)
Comment (vadim-moz) [inline src/tokenizers/strspn.h]: done

File: src/tokenizers_test.cpp (hunk 3 above)
Comment (vadim-moz) [inline src/tokenizers_test.cpp]: done

File: src/tokenizers_test.cpp (hunk 2 above)
Comment (vadim-moz) [inline src/tokenizers_test.cpp]: done

**Comment by vadim-moz** [comment]
PTAL

**Comment by b4hand** [comment]
LGTM

**Comment by tammybailey** [comment]
lgtm

**Comment by vadim-moz** [comment]
Thank you!

**Pull request merged**